plugin/main/src/kotlinx/benchmark/gradle/SuiteSourceGenerator.kt

```diff
@@ -99,21 +100,99 @@ class SuiteSourceGenerator(val title: String, val module: ModuleDescriptor, val
                 .filter { it.annotations.any { it.fqName.toString() == stateAnnotationFQN } }
-                .filter { it.modality != Modality.ABSTRACT }
-                .forEach {
-                    generateBenchmark(it)
+
+            // Contains both directly and indirectly annotated ClassDescriptors.
+            val annotatedInheritableClassDescriptors = inheritableClassDescriptors
+                .filter { inheritableClass ->
+                    directlyAnnotatedInheritableClassDescriptors.forEach { annotatedInheritableClass ->
+                        if (inheritableClass.isSubclassOf(annotatedInheritableClass)) {
+                            // If any benchmark class is not annotated but extended with an abstract class
+                            // annotated with @State, it should be included when generating benchmarks.
+                            return@filter true
+                        }
+                    }
+                    return@filter false
                 }
+
+            val annotatedClassDescriptors = mutableListOf<AnnotatedClassDescriptor>()
+                .apply {
+                    classDescriptors
+                        .filter { it.modality != Modality.ABSTRACT }
+                        .forEach {
+                            if (it.annotations.any {
+                                annotationDescriptor -> annotationDescriptor.fqName.toString() == stateAnnotationFQN
+                            }) {
+                                add(AnnotatedClassDescriptor(it))
+                            } else {
+                                val parent = it.getParentAnnotated(annotatedInheritableClassDescriptors)
+                                if (parent != null) {
+                                    add(AnnotatedClassDescriptor(it, parent))
+                                }
+                            }
+                        }
+                }
+
+            annotatedClassDescriptors.forEach {
+                generateBenchmark(it)
+            }
         }
 
         for (subpackageName in module.getSubPackagesOf(packageView.fqName, MemberScope.ALL_NAME_FILTER)) {
             processPackage(module, module.getPackage(subpackageName))
         }
     }
 
-    private fun generateBenchmark(original: ClassDescriptor) {
+    /** @param annotatedInheritableClassDescriptors descriptors of classes which are inheritable and directly annotated.
+     * with [stateAnnotationFQN].
+     * @return Top most parent class descriptor which was annotated with [stateAnnotationFQN] and inherited by `this` class
+     * descriptor or null if none of its parent classes are not annotated.*/
+    private fun ClassDescriptor.getParentAnnotated(
+        annotatedInheritableClassDescriptors: List<ClassDescriptor>
+    ): ClassDescriptor? {
+        annotatedInheritableClassDescriptors.forEach { annotatedAbstractClass ->
+            if (this.isSubclassOf(annotatedAbstractClass)) {
```

Comment:
Could it be that there are multiple superclasses for `this` class descriptor?

Reply:
I was expecting this! So as for phase-1 I was just looking out for bare bones implementation which means this here just a base setup. Currently this function picks the **first** superclass that is directly annotated. Now there are some cases:
1) Simple straight forward: Only one annotated superclass.
2) Multiple superclasses but only one is annotated.
3) Multiple superclasses and intermediaries (super class of this but sub class of the top most annotated parent class) are annotated as well.
4) Multiple superclasses and intermediaries with annotations E.x., `@Warmup`, `@Measurement` etc without `@State`.

What should be the route to follow here?
For 4th point, my suggestion is that independent annotations should be ignored. For the 3rd point, I would suggest nearest annotated parent. 1st and 2nd are easily dealt with.

Reply:
> What should be the route to follow here?

It depends on how JMH handles this aspect. Being consistent with that library is important, since it runs the benchmarks on the JVM. Achieving consistent behavior across other platforms would be ideal. It is worth mentioning that we do not aim to replicate all of JMH's sophistication, but when something is not supported in non-JVM targets, we would like to issue a warning (or throw an exception) about it.

Reply:
I would greatly appreciate it if you could check how JMH handles this case. It would greatly help us decide which route we should take.